File: python/utils/network.py
import torch.nn as nn


class Network(nn.Sequential):
    def __init__(self, input_dim: int,
                 output_dim: int,
                 units=16,
                 hidden_layers=1,
                 ):
        super(Network, self).__init__()
        assert hidden_layers > 0 and isinstance(hidden_layers, int), \
               "Layers should be an positive int"
        assert units > 0 and isinstance(units, int), \
               "Units should be an positive int"
        self.input_dim = input_dim
        self.hidden_layers = hidden_layers
        self.model = nn.Sequential(
                nn.Linear(input_dim, units),
                nn.ReLU()
                )
        if hidden_layers > 1:
            for i in range(hidden_layers):
                if i < hidden_layers//2:
                    input_units = 2**i
                    output_units = 2**(i+1)
                elif hidden_layers % 2 == 1 and i == hidden_layers//2:
                    input_units = 2**i
                    output_units = input_units
                elif i >= hidden_layers//2:
                    input_units = 2**(hidden_layers-i)
                    output_units = 2**(hidden_layers-i-1)
                else:
                    raise Exception("Unknown error")
                input_units *= units
                output_units *= units
                self.model.append(nn.Linear(input_units, output_units))
                self.model.append(nn.ReLU())
        self.model.append(nn.Linear(units, output_units))

File: rust/src/utils/network.rs
use tch::nn;

pub fn network(vs: &nn::Path, input_dim: i64, output_dim: i64, units: i64, hidden_layers: u32) -> nn::Sequential{
    let hidden_layers = hidden_layers as u32;
    if hidden_layers <= 0{ 
        panic!("Layers should be positive");
    }
    let mut seq = nn::seq();
    seq = seq.add(nn::linear(vs / "input_layer", 
            input_dim,
            units,
            Default::default(),
    ))
        .add_fn(|xs| xs.relu());

    if hidden_layers > 1 {
        let mut input_units;
        let mut output_units;

        for layers in 0..hidden_layers {            
            if layers < hidden_layers/2 {
                input_units = i64::pow(2,layers); 
                output_units = i64::pow(2,layers+1);
            } 
            else if hidden_layers%2 == 1 && layers == hidden_layers/2{
                input_units = i64::pow(2,layers); 
                output_units = input_units; 
            }
            else if layers >= hidden_layers/2{
                input_units = i64::pow(2,hidden_layers-layers); 
                output_units = i64::pow(2,hidden_layers-layers-1); 
            }
            else {
                panic!("Unknown error");
            }
            seq = seq.add(nn::linear(vs, 
                    input_units*units,
                    output_units*units, 
                    Default::default()))
                .add_fn(|xs| xs.relu());
            }
    }
    seq.add(nn::linear(vs / "output_layer", units, output_dim, Default::default()))
}

